### PR TITLE
no_std for algebra and r1cs-core

### PR DESCRIPTION
**Pull request by debris**
Hey! This is an ongoing attempt to make zexe `no_std` compatible. 

- `algebra` compiles with `no_std`
  - removed `byteorder` dependency
  - uses `algebra/src/io.rs` as an `std::io` counterpart
  - uses `algebra/serialize/Error` as `std::error::Error` trait counterpart
  - uses `rand::test_rng()` for generating random numbers in tests
  - split `algebra-benches` into it's own crate - [reason](https://github.com/scipr-lab/zexe/pull/76#issuecomment-579924945)
- `r1cs-core` compiles with `no_std`
- ~`ff-fft` compiles with `no_std`~
  - I did a bunch of changes to `ff-fft`, but eventually decided to postpone `no-std` support for this library till the next PR. This PR is already big enough
- removed floating point operations (#79)
- removed `println!` from non-test code
- added `no-std` bare metal build for `algebra`, `r1cs-core` and ~`ff-fft`~ to `.travis.yml`
- updated `smallvec` to `1.2`, because it contains a fix for [this issue](https://github.com/servo/rust-smallvec/issues/197)

### CODE REVIEW

**Comment by Pratyush** [comment]
Hey! Awesome! This would indeed be very useful!

Btw, you don't need to remove byteorder; you can use it with `default-features = false` to have it run in `no_std` mode.

File: algebra/src/lib.rs [hunk 1]
```diff
@@ -33,6 +34,20 @@
 )]
 #![forbid(unsafe_code)]
 
+#[cfg(not(feature = "std"))]
+macro_rules! println {
```
Comment (Pratyush) [inline algebra/src/lib.rs]: Do we use this macro anywhere in the crate? If we do, it's probably a bug.

File: algebra/src/lib.rs [hunk 2]
```diff
@@ -82,3 +97,5 @@ pub mod prelude {
 
     pub use num_traits::{One, Zero};
 }
+
+mod fake_io;
```
Comment (Pratyush) [inline algebra/src/lib.rs]: Can we rename this to `io`, and make it conditionally equal the `std` module:
```suggestion
#[cfg(not(feature = "std"))]
mod io;
#[cfg(feature = "std")]
pub use std::io;
```

File: algebra/src/serialize/error.rs [hunk 3]
```diff
@@ -14,26 +15,26 @@ pub enum SerializationError {
...
-impl Error for SerializationError {
-    fn source(&self) -> Option<&(dyn Error + 'static)> {
-        None
-    }
-}
+//impl Error for SerializationError {
+    //fn source(&self) -> Option<&(dyn Error + 'static)> {
+        //None
+    //}
+//}
```
Comment (Pratyush) [inline algebra/src/serialize/error.rs]: Maybe instead of commenting these out, replace it with a `#[cfg(feature = "std")]`?

File: algebra/src/serialize/error.rs [hunk 4]
```diff
@@ -14,26 +15,26 @@ pub enum SerializationError {
...
 }
 
 impl fmt::Display for SerializationError {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
-        if let SerializationError::IoError(e) = self {
-            write!(f, "I/O error: ")?;
-            e.fmt(f)
+        if let SerializationError::IoError(_) = self {
+            write!(f, "I/O error")
+            //e.fmt(f)
```
Comment (Pratyush) [inline algebra/src/serialize/error.rs]: we can just delete this line instead of commenting it out.

File: algebra/src/to_field_vec.rs [hunk 5]
```diff
@@ -6,9 +8,10 @@ use crate::{
         ProjectiveCurve,
     },
     Fp2, Fp2Parameters, FpParameters, Field, PrimeField,
+    serialize::SerializationError,
 };
 
-type Error = Box<dyn std::error::Error>;
+type Error = Box<SerializationError>;
```
Comment (Pratyush) [inline algebra/src/to_field_vec.rs]: We can just replace this with `SerializationError` now, right? No need for the `Box`?

File: algebra/src/bytes.rs [hunk 6]
```diff
@@ -282,17 +282,18 @@ impl<'a, T: 'a + ToBytes> ToBytes for &'a T {
...
-    }
-}
+//impl FromBytes for Vec<u8> {
+    //#[inline]
+    //fn read<R: Read>(mut reader: R) -> IoResult<Self> {
+        //let mut buf = Vec::new();
+        //let _ = reader.read_to_end(&mut buf)?;
+        //Ok(buf)
+    //}
+//}
```
Comment (Pratyush) [inline algebra/src/bytes.rs]: We don't need to delete this, right? It should be available even in `no_std` because we're importing `alloc`?

**Review by Pratyush — COMMENTED** [review]
Added some comments on the changes to `algebra`

File: algebra/src/lib.rs (hunk 1 above)
Comment (debris) [reply algebra/src/lib.rs]: Yes, a bunch of tests uses it. Should I replace it with a `log` crate?

File: algebra/src/lib.rs (hunk 2 above)
Comment (debris) [reply algebra/src/lib.rs]: yes, I think it's doable

File: algebra/src/serialize/error.rs (hunk 3 above)
Comment (debris) [reply algebra/src/serialize/error.rs]: I was thinking of removing it completely, but of course, I can add `#[cfg(feature = "std")]` for backwards compatibility

File: algebra/src/serialize/error.rs (hunk 4 above)
Comment (debris) [reply algebra/src/serialize/error.rs]: yes, the pr is not finished yet 😅  I wouldn't commit in commented out code :)

File: algebra/src/to_field_vec.rs (hunk 5 above)
Comment (debris) [reply algebra/src/to_field_vec.rs]: Sure. I didn't remove the `Box`, cause I thought that it may be here for performance reasons (e.g. to avoid allocating space on stack for a big `Error`), but I don't think it is a case here

Comment (Pratyush) [reply algebra/src/to_field_vec.rs]: Yeah, earlier I didn't want to do the work of defining a custom Error enum, so I took the lazy shortcut =P

File: algebra/src/lib.rs (hunk 2 above)
Comment (debris) [reply algebra/src/lib.rs]: done

File: algebra/src/to_field_vec.rs (hunk 5 above)
Comment (debris) [reply algebra/src/to_field_vec.rs]: done :)

File: algebra/src/lib.rs (hunk 1 above)
Comment (Pratyush) [reply algebra/src/lib.rs]: hmm maybe you can add
```rust
#[cfg(and(test, not(feature = "std")))]
#[macro_use]
extern crate std;
```

File: algebra/src/bytes.rs (hunk 6 above)
Comment (debris) [reply algebra/src/bytes.rs]: I removed it cause it's not used anywhere. I could re-add it, but that would imply that we have to add `read_to_end` to `algebra::io::Read`

**Comment by debris** [comment]
I think the PR is ready. Once merged, I'll work on adding support to the remaining libraries.

> Btw, you don't need to remove byteorder; you can use it with default-features = false to have it run in no_std mode.

Since rust 1.32, `byteorder` is not really needed, cause all unsigned integer primitives have [`from_le_bytes`](https://doc.rust-lang.org/std/primitive.u64.html#method.from_le_bytes) method

**Comment by kobigurk** [comment]
Naive question since I haven't looked closely - does this allow verifying
proofs on wasm?

**Comment by debris** [comment]
@kobigurk not yet. I have another branch that I haven't pushed yet (it's not finished) that allows to do that

**Comment by debris** [comment]
update*
I pushed the branch here: https://github.com/scipr-lab/zexe/compare/master...debris:groth16-prover?expand=1

**Comment by debris** [comment]
I did some tests, and I need to work on this PR a little bit more. It's because of how `cargo` build system works. 

`no-std` version of `algebra` crate depends on `no-std` version of `num-traits`. Unfortunately, one of `dev-dependencies` of `algebra` is `criterion` which depends on `std` version of `num-traits`. Because of that, `cargo`  builds `num-traits` only once - with `std` and links exactly this version of `num-traits` to `algebra`.

this build issue applies also to several other dependencies

I believe the only way to solve this problem is to separate `algebra` and `algebra-benches` from each other

**Comment by Pratyush** [comment]
Yeah the unification of features across `dev-dependencies` is *very* annoying. We can also try submitting a PR to upstream `criterion` to make them use `num-traits` in `no_std` mode

**Comment by Pratyush** [comment]
Ok I made the PR, but it might make sense to have a separate benches crate anyway, just to avoid future fragility.

File: algebra/Cargo.toml [hunk 7]
```diff
@@ -37,4 +36,5 @@ criterion = "0.3.1"
 rand_xorshift = { version = "0.2" }
 
 [features]
+std = []
 parallel = [ "rayon" ]
```
Comment (Pratyush) [inline algebra/Cargo.toml]:
```suggestion
parallel = [ "rayon", "std" ]
```

File: groth16/Cargo.toml [hunk 8]
```diff
@@ -23,7 +23,7 @@ edition = "2018"
 ################################# Dependencies ################################
 
 [dependencies]
-algebra = { path = "../algebra", features = [ "parallel" ] }
+algebra = { path = "../algebra", features = [ "parallel", "std" ] }
```
Comment (Pratyush) [inline groth16/Cargo.toml]:
```suggestion
algebra = { path = "../algebra", features = [ "parallel" ] }
```

Comment (Pratyush) [reply groth16/Cargo.toml]: Though, if you want these crates to also be `no_std`, you probably need to drop the `parallel` feature.

I'd recommend adding a `parallel` feature to this crate, and then making that activate `algebra/parallel`.

**Review by Pratyush — CHANGES_REQUESTED** [review]
Left some comments. Also, to make `ff-fft` `no_std`, you probably need to have a conditional `parallel` feature that activates the parallelization infrastructure https://github.com/scipr-lab/zexe/blob/master/ff-fft/src/multicore.rs

When `parallel` is off the `scope` method would presumably forward to just `chunks` or something. Maybe something similar to https://github.com/matter-labs/bellman/blob/ff783cb207b4d61b8444a4aad91151ee498ccb32/src/singlecore.rs#L56

File: groth16/Cargo.toml (hunk 8 above)
Comment (Pratyush) [reply groth16/Cargo.toml]: (same for `gm17`)

Comment (debris) [reply groth16/Cargo.toml]: I'll work on this in the next PR

**Comment by debris** [comment]
> Left some comments. Also, to make ff-fft no_std, you probably need to have a conditional parallel feature that activates the parallelization infrastructure https://github.com/scipr-lab/zexe/blob/master/ff-fft/src/multicore.rs
> 
> When parallel is off the scope method would presumably forward to just chunks or something. Maybe something similar to https://github.com/matter-labs/bellman/blob/ff783cb207b4d61b8444a4aad91151ee498ccb32/src/singlecore.rs#L56

I decided to postpone making `ff-fft` `no-std` until the next PR. This one is already big enough. And I think it's ready for another review :)

File: algebra/src/rand.rs [hunk 9]
```diff
@@ -16,3 +16,14 @@ where
         rng.sample(Standard)
     }
 }
+
+pub fn test_rng() -> rand::rngs::StdRng {
```
Comment (Pratyush) [inline algebra/src/rand.rs:21]: ~We can put this under `#![cfg(test)]` so that it's not exported.~

I see that it's used in `ff-fft` and other crates. Maybe we can add a line of documentation indicating that this should be used only for testing purposes, and not for any real-world usage?

File: r1cs-core/src/impl_lc.rs [hunk 10]
```diff
@@ -13,14 +13,14 @@ impl<F: Field> AsRef<[(Variable, F)]> for LinearCombination<F> {
...
-        LinearCombination(smallvec![(var, coeff)])
+        LinearCombination(smallvec!((var, coeff)))
     }
 }
 
 impl<F: Field> From<Variable> for LinearCombination<F> {
     #[inline]
     fn from(var: Variable) -> Self {
-        LinearCombination(smallvec![(var, F::one())])
+        LinearCombination(smallvec!((var, F::one())))
```
Comment (Pratyush) [inline r1cs-core/src/impl_lc.rs]: Could you undo the change from `smallvec![...]`? It's supposed to be used similarly to the `vec![...]` macro

File: algebra-benches/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+description = "A benchmark library for finite fields and elliptic curves"
+homepage = "https://libzexe.org"
+repository = "https://github.com/scipr/zexe"
+documentation = "https://docs.rs/algebra/"
+keywords = ["cryptography", "finite fields", "elliptic curves", "pairing"]
+categories = ["cryptography"]
+include = ["Cargo.toml", "src", "README.md", "LICENSE-APACHE", "LICENSE-MIT"]
+license = "MIT/Apache-2.0"
+edition = "2018"
+
```
Comment (Pratyush) [inline algebra-benches/Cargo.toml:22]: Can we add a `publish = false` here so that it isn't accidentally published to `crates.io`, and it's clear that it's meant only for local benchmarking?

**Review by Pratyush — CHANGES_REQUESTED** [review]
Awesome! The PR looks mostly good to me. CI is failing because (I think because `r1cs-std` doesn't yet use `r1cs-core`/`algebra` under the `std` flag).

Oh and if possible, could you squash the commits into logically related ones?

**Comment by debris** [comment]
> Awesome! The PR looks mostly good to me. CI is failing because (I think because `r1cs-std` doesn't yet use `r1cs-core`/`algebra` under the `std` flag).

I see... To fix the problem, I decided to make `algebra` and `r1cs-core` `std` by default. It's an approach that many other projects also have (e.g. `serde`). To compile it for `no-std`, one needs to set `default-features = false`

> Oh and if possible, could you squash the commits into logically related ones?

Hmm... is it possible to just squash all of them into a single commit using github's `Squash and merge` option? 😅

**Comment by debris** [comment]
CI is green 🎉

**Comment by Pratyush** [comment]
Awesome, merged! Thank you so much for the PR!
